o Better implementation of auto-profiling:

jmrosinski opened the pull request:
- No more need for post-processing script to turn addresses into names
  - Use libunwind if available in GPTLstart/GPTLstop
  - Use backtrace if libunwind not available
  - No more need for configure arg to test auto-profiling: test it if either
    libunwind or bactrace are found
o Fix OpenMP testing:
  - Default is "enabled"
  - Set proper ifdef
o Fix pthread testing:
  - Default is "disabled"
  - Set proper ifdef
  - Only test if OpenMP is NOT enabled
o Look for PAPI library only if --enable-papi specified

jmrosinski committed:
o Better implementation of auto-profiling:
  - No more need for post-processing script to turn addresses into names
  - Use libunwind if available in GPTLstart/GPTLstop
  - Use backtrace if libunwind not available
  - No more need for configure arg to test auto-profiling: test it if either
    libunwind or bactrace are found
o Fix OpenMP testing:
  - Default is "enabled"
  - Set proper ifdef
o Fix pthread testing:
  - Default is "disabled"
  - Set proper ifdef
  - Only test if OpenMP is NOT enabled
o Look for PAPI library only if --enable-papi specified

edhartnett commented on configure.ac:
```diff
@@ -42,16 +42,48 @@ if test "x$have_nanotime" = xyes; then
+    AC_MSG_RESULT([$enable_pthread])
+  ])
+  
+])
```
Things work a lot better if you check all enables/disables unconditionally. That is, don't have an option that only appears if another option has been specified, it's too confusing and hard to maintain as it grows.

So evaluate the --enable-pthread even if useomp was not specified.

Then, after all --enables/disables are processed, you can use some simple shell if statements to error out if incompatible options are enabled.

For each option, it's helpful if there is always a shell variable defined to either yes or no. For example:

```
# Does the user want to disable pnetcdf?
AC_MSG_CHECKING([whether pnetcdf is to be used])
AC_ARG_ENABLE([pnetcdf],
              [AS_HELP_STRING([--disable-pnetcdf],
                              [Disable pnetcdf use.])])
test "x$enable_pnetcdf" = xno || enable_pnetcdf=yes
AC_MSG_RESULT([$enable_pnetcdf])
AM_CONDITIONAL(BUILD_PNETCDF, [test "x$enable_pnetcdf" = xyes])

```
After this code is run, $enable_pnetcdf is set to "yes" or "no". So later in the script I can always do:

```
if test "x$enable_pnetcdf" = xyes; then
   # Do some stuff for pnetcdf builds.
fi

```

You do not need to use the AS_IF macro.

edhartnett commented on configure.ac:
```diff
@@ -65,17 +97,6 @@ fi
-              [test autoprofileing])])
-test "x$enable_autoprofile_testing" = xyes || enable_autoprofile_testing=no
-AM_CONDITIONAL(TEST_AUTOPROFILE, [test $enable_autoprofile_testing = yes])
-AC_MSG_RESULT($enable_autoprofile_testing)
```
You are just removing this. Is that your intention?

edhartnett commented on configure.ac:
```diff
@@ -88,22 +109,32 @@ else
+              [enable papi])])
+test "x$enable_papi" = xyes || enable_papi=no
+AC_MSG_RESULT($enable_papi)
+
```
I usually do all my ARG processing (enables/disables) at the top of the script, the all my $enable_whatever variables are set to "yes" or "no" and the rest of the script I can use them to make decisions.

edhartnett commented on test/Makefile.am:
```diff
@@ -4,6 +4,9 @@
 AM_CPPFLAGS = -I$(top_srcdir)/include
 
+# Need optimization disabled for correct behavior of some tests
+CFLAGS += -g -O0
```
We must not set CFLAGS. There is an AM_CFLAGS you can use instead, if you must.

But instead, consider erroring out in configure if optimization behavior is not correct, instead of guessing flags.

edhartnett commented on test/Makefile.am:
```diff
@@ -14,6 +17,26 @@ noinst_PROGRAMS = gran_overhead printwhileon
 
+# Test autoprofiling if either HAVE_LIBUNWIND or HAVE_BACKTRACE
+if HAVE_LIBUNWIND
+  AM_LDFLAGS += -rdynamic -lunwind
```
You may mean cygprofile_LDFLAGS here instead of AM_LDFLAGS.

AM_LDFLAGS will all these flags for all compiles in this directory. cygprofile_LDFLAGS will just add it for that one program.

jmrosinski commented:
Thanks for the feedback. Regarding your comments:

o Agree it's cleaner to check *all* ARG_ENABLE settings, then script logic
later for incompatible settings. I'll do that.

o Yes AS_IF seemed unnecessary. I assumed it was useful since there are web
pages explaining its use.

o Right there's no longer need for user input regarding TEST_AUTOPROFILE.
WIth the new scheme, auto-profiling (that's the cygprofile stuff) will be
tested if either libunwind or backtrace are available. If neither are
available it will not be tested. No user input required.

o Unfortunately for testing some things, auto-profiling is one of them, it
often fails with high optimization. So the code in "test/" needs different
flags than the library itself. Yes I'll change the AM_CFLAGS and
cygprofile_LDFLAGS per your suggestions.

Thanks,
Jim


On Mon, Apr 1, 2019 at 9:34 AM Ed Hartnett <notifications@github.com> wrote:

> *@edhartnett* commented on this pull request.
> ------------------------------
>
> In test/Makefile.am
> <https://github.com/jmrosinski/GPTL/pull/48#discussion_r270927487>:
>
> > @@ -14,6 +17,26 @@ noinst_PROGRAMS = gran_overhead printwhileon
>  check_PROGRAMS = tst_simple global
>  TESTS = tst_simple global
>
> +# Test autoprofiling if either HAVE_LIBUNWIND or HAVE_BACKTRACE
> +if HAVE_LIBUNWIND
> +  AM_LDFLAGS += -rdynamic -lunwind
>
> You may mean cygprofile_LDFLAGS here instead of AM_LDFLAGS.
>
> AM_LDFLAGS will all these flags for all compiles in this directory.
> cygprofile_LDFLAGS will just add it for that one program.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmrosinski/GPTL/pull/48#pullrequestreview-221188286>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN7k11KDVWrRY7P4cmfOFbRoiq6Ag5ks5vcicKgaJpZM4cV-qJ>
> .
>

jmrosinski committed:
o More bugfixes
o Employed library hack to compile cygprofilesubs differently

jmrosinski commented:
I absolutely CANNOT get this to work correctly using anything but CFLAGS.
AM_CFLAGS seems to be ignored, and other things put the flags BEFORE the
default flags, so the defaults (e.g. -O2) override what I want. The overall
issue is this:

The GPTL library should be compiled with high optimization because we want
as little intrusion on user timings as possible. But, for auto-profiling
(one of GPTL's most valuable capabilities), the *application *code (not the
library itself), needs to be compiled with low optimization (no higher than
-O1) to work correctly. Therefore code in the test/ directory needs to be
built with lower optimization than the library itself.

Do you have a solution in mind? Why is it bad to modify CFLAGS? If you
wouldn't mind having a look, I've committed/pushed the latest working code
to the branch *feature/autoprofile* (modifies CFLAGS in test/ because
that's all I could get to work). Also if some of check_PROGRAMS, TESTS, and
noinst_PROGRAMS can be eliminated that would be good--I just put them all
in there not knowing what they meant trying to get something to work.

If you'd like to test the code, first you'll need libunwind installed (I
think I just did apt install libunwind-dev). Then after autoreconf -i:

./configure --enable-autoprofile
make clean; make
cd test; ./cygprofile

If cygprofile complains about anything it's probably due to wrong flags as
indicated above.

Happy to try a better fix if you have one in mind.

Thanks,
Jim






On Mon, Apr 1, 2019 at 9:33 AM Ed Hartnett <notifications@github.com> wrote:

> *@edhartnett* commented on this pull request.
> ------------------------------
>
> In test/Makefile.am
> <https://github.com/jmrosinski/GPTL/pull/48#discussion_r270926733>:
>
> > @@ -4,6 +4,9 @@
>  # Find include directory.
>  AM_CPPFLAGS = -I$(top_srcdir)/include
>
> +# Need optimization disabled for correct behavior of some tests
> +CFLAGS += -g -O0
>
> We must not set CFLAGS. There is an AM_CFLAGS you can use instead, if you
> must.
>
> But instead, consider erroring out in configure if optimization behavior
> is not correct, instead of guessing flags.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmrosinski/GPTL/pull/48#pullrequestreview-221187297>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN7g-MlsIIOQB7gucWy4SgVXgV6Frcks5vciavgaJpZM4cV-qJ>
> .
>

edhartnett commented on test/Makefile.am:
```diff
@@ -19,21 +19,29 @@ TESTS = tst_simple global
+  AM_LDFLAGS        += -rdynamic -lunwind
+  check_PROGRAMS    += cygprofile
+  TESTS             += cygprofile
+  noinst_PROGRAMS   += cygprofile
```
You don't have to list it as a noinst_PROGRAMS if it is a check_PROGRAMS. check programs are already not installed.

jmrosinski committed:
o Cleanup configure output and rewrite for less indentation
  of AC_ARG_ENABLE invocations

edhartnett commented on configure.ac:
```diff
@@ -62,28 +62,33 @@ AS_IF([test "x$enable_openmp" != xno], [
+AC_ARG_ENABLE([pthread],
+    AS_HELP_STRING([--enable-pthread], [Enable pthread capability]))
+
+if test "x$enable_pthread" = xyes; then
```
At this point $enable_pthread will be unset if the user accepts the default. This leads to confusing errors. What I have done is settled on this idiom, which I would like to use in the GPTL configure.ac:

```
# Does the user want to enable logging?
AC_MSG_CHECKING([whether debug logging is enabled])
AC_ARG_ENABLE([logging],
              [AS_HELP_STRING([--enable-logging],
                              [enable debug logging capability (will negatively impact performance). \
			      This debugging feature is probably only of interest to PIO developers.])])
test "x$enable_logging" = xyes || enable_logging=no
AC_MSG_RESULT([$enable_logging])
```

Note that after this code is run, $enable_logging is always defined, to either "yes" or "no". This is useful, for example in the AC_MSG_RESULT call, which provides valuable feedback in the configure output, to help with support.

The way this has to be done changes a little when the features is enabled by default, instead of disabled by defualt. Here's one of those:

```
# Does the user want to disable pnetcdf?
AC_MSG_CHECKING([whether pnetcdf is to be used])
AC_ARG_ENABLE([pnetcdf],
              [AS_HELP_STRING([--disable-pnetcdf],
                              [Disable pnetcdf use.])])
test "x$enable_pnetcdf" = xno || enable_pnetcdf=yes
AC_MSG_RESULT([$enable_pnetcdf])

```

(Note that the AC_ARG_ENABLE call indicates which is the default by providing the non-default option.)

jmrosinski committed:
Semi-functional auto-profiling. Still some issues with C++ need resolution

o Bugfix for __cyg* routines where they need to return immediately on error
o Add --enable-debug to configure.ac
o Reinstall --enable-unwind
o Add MAX_SYMBOL_NAME for mangled C++
o Reinstall getentry_instr_sim
o Add print_callstack debug capability
o Remove check_PROGRAMS from Makefile.am per Ed Hartnett
o Remove some GPTL calls from cygprofile.c which will no longer work due to
  inability to recover some info from auto-instrumented codes

jmrosinski committed:
o Bugfix: Add OPENMP_CFLAGS to CFLAGS when OpenMP enabled
o Reformatting to shrink width of print:
  - Mainly helps auto-profiled output with C++ mangled names which can be long
  - Indenting computation now computes longest per-timer max (depth plus
    number of chars) rather than max (depth) plus max (number of chars)
  - Right-justify everything
  - Almost everything is 9 chars
o More diagnostics when imperfect_nest = true
o Remove some warnings
o Remove #pragma unroll
o Add omptest.c, imperfect_nest.c to tests

jmrosinski committed:
o Unified API for MPI enabled vs. disabled. 1 extra arg for MPI disabled, no
  change for MPI enabled.

o Some minor mods for preparation for moving to C++

edhartnett commented on include/gptl.h:
```diff
@@ -6,6 +6,14 @@
 #define GPTL_H
 
+// mpi.h needed for MPI_Comm type
+// When MPI not enabled, use an int as a placeholder
```
Don't use C++ comments, use C comments.

jmrosinski committed:
o Added configure to branch because autoreconf doesn't seem to work properly
  on Mac.

jmrosinski commented:
I would agree, except right now I am working on migrating GPTL to C++. I
want to use things like reference variables, and namespaces, to clean up
the code. Though right now I can't even compile with g++ as it doesn't like
things like "char *p = zzz".

Jim


On Mon, Apr 8, 2019 at 2:21 PM Ed Hartnett <notifications@github.com> wrote:

> *@edhartnett* commented on this pull request.
> ------------------------------
>
> In include/gptl.h
> <https://github.com/jmrosinski/GPTL/pull/48#discussion_r273222059>:
>
> > @@ -6,6 +6,14 @@
>  #ifndef GPTL_H
>  #define GPTL_H
>
> +// mpi.h needed for MPI_Comm type
> +// When MPI not enabled, use an int as a placeholder
>
> Don't use C++ comments, use C comments.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmrosinski/GPTL/pull/48#pullrequestreview-224066075>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN7vG0w-q6-YSvLuB5gRR8pBwjE9vSks5ve6TXgaJpZM4cV-qJ>
> .
>

jmrosinski committed:
o Something didn't work right with the test for OMP in configure, so simplify
  configure.ac to always check, and enable/disable based on the results

jmrosinski committed:
o Change auto-profiling default from libunwind to backtrace based on JEDI
  and Mac

jmrosinski committed:
o Fix openmp tests for OMP not available
o Fix clang complaints

jmrosinski committed:
Merge branch 'feature/autoprofile' of https://github.com/jmrosinski/GPTL into feature/autoprofile

jmrosinski committed:
o Cleanup configure.ac (no functionality changes0
o Change default underlying timer (UTR) to nanotime on x86

jmrosinski committed:
o Add "_experimental" to version
o Add gptl.inc to files to be installed

jmrosinski committed:
Merge remote-tracking branch 'origin/master' into feature/autoprofile

edhartnett commented:
I suggest you merge this. It's not a good idea to keep PRs open for a long time. Each PR should be a targeted change, linked to an issue.

This PR is already very large and complex - too much so for easy review. If you want the benefits of review, you must present your work in bite-sized pieces.

jmrosinski commented:
Yes I agree and will merge it soon. I goofed and issued the PR too soon as there turned out to be issues that still needed to be addressed. The volume of changes is large, but that is the nature of migrating from post-processing autoprofiling to doing it inline.

jmrosinski committed:
Merge remote-tracking branch 'origin/master' into feature/autoprofile

jmrosinski committed:
Merge remote-tracking branch 'origin/master' into feature/autoprofile
